### ISSUE TITLE
https://vidlox.me/ Tampermonkey sandbox preparation failed. This usually is caused by a third-party extension. ReferenceError: xxxx

### ISSUE THREAD

**Issue by raszpl** [Tampermonkey/tampermonkey#805]
## Expected Behavior
Reliable script injection on https://vidlox.me/ https://www.vidlox.me/

## Actual Behavior
_randomly_ https://vidlox.me/ results in
> Tampermonkey sandbox preparation failed. This usually is caused by a third-party extension. ReferenceError: somerandomstring

and no script execution in Vivaldi browser, works correctly about ~4 out of 10 reloads.
or no script execution without any Console errors/information in Opera, works correctly about ~1 out of 10 reloads.

About 10 extensions in Vivaldi. Only Tampermonkey/uBlock Origin/uBlock Origin Extra in Opera.

## Specifications
- Vivaldi/2.8.1664.40 (Chrome/77.0.3865.93)
- OPR/64.0.3417.83 (Chrome/77.0.3865.120)
- TM: v4.8.5890, v4.9.6079 
- OS: W10

## Script

> // ==UserScript==
> // @name         vidlox
> // @match        https://www.vidlox.me/*
> // @match        https://vidlox.me/*
> // @grant        none
> // ==/UserScript==
> 
> console.log("hello");
> alert("worked");

**Comment by derjanb** [comment]
Can not reproduce. Is this still an issue?

**Comment by derjanb** [comment]
If so, can you please re-open the issue and send the full error message? Thanks.

**Comment by raszpl** [comment]
Opera 64.0.3417.83

```
VM31:1 Tampermonkey sandbox preparation failed. This usually is caused by a third-party extension. ReferenceError: yo6kpep
    at abort (<anonymous>:68:15)
    at <anonymous>:8:323
    at Array.forEach (<anonymous>)
    at <anonymous>:8:296
    at <anonymous>:9:88
    at <anonymous>:11:396
```


Vivaldi | 2.9.1735.3

```
Tampermonkey sandbox preparation failed. This usually is caused by a third-party extension. ReferenceError: zmrmpur
    at abort (<anonymous>:68:15)
    at <anonymous>:8:308
    at Array.forEach (<anonymous>)
    at <anonymous>:8:287
    at <anonymous>:9:88
    at <anonymous>:11:396
```

I dont have pure Chrome. Its _random_, but triggers at least couple of times per 10 page reloads on https://vidlox.me/. Dont know how to reopen bug reports

**Comment by raszpl** [comment]
Its definitely a conflict with 'uBlock Origin 1.23.0'. Disabling uBO fixes it permanently. Randomness might be due to random order extensions are loaded/initialized? Still no idea why it only happens on this one particular website. It even happens on 404 page (going to https://vidlox.me/somerandomstring)
I tried Chrome devtool request blocking all external JS scripts loaded by this website, no change.
Tried disabling JS execution using uBO "Click to disable JavaScript on this site", no change.
Tried saving vidlox.me.html to disk and re-hosting locally using
`python -m SimpleHTTPServer 8000`
http://localhost:8000/vidlox.me.html hosted version doesnt trigger this bug, tempermonkey works 100% reliably this way.

Tried inject mode **instant**, still bugging in same random way, only change is additional warning:
```
content.js:formatted:745 [Deprecation] Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.
a @ content.js:formatted:745
(anonymous) @ content.js:formatted:763
(anonymous) @ content.js:formatted:758
(anonymous) @ content.js:formatted:760
(anonymous) @ content.js:formatted:20
b @ content.js:formatted:13
c @ content.js:formatted:25
(anonymous) @ page.js:1
VM1555:1 Tampermonkey sandbox preparation failed. This usually is caused by a third-party extension. ReferenceError: xponjv7
    at abort (<anonymous>:68:15)
    at <anonymous>:8:323
    at Array.forEach (<anonymous>)
    at <anonymous>:8:296
    at <anonymous>:9:88
    at <anonymous>:11:396
```

**Comment by raszpl** [comment]
Found another website with same issue: https://www.shameless.com/dmca.php (NSFW safe subpage, main address is not). Again only 2 extensions loaded  'uBlock Origin 1.23.0' and TM (4.10.6106).

**Comment by raszpl** [comment]
and another one, https://mixdrop.co/copyright/

```
Tampermonkey sandbox preparation failed. This usually is caused by a third-party extension. ReferenceError: ugngsq5
    at validate (<anonymous>:94:19)
    at Function.get (<anonymous>:99:13)
    at <anonymous>:6:193
    at <anonymous>:9:88
    at <anonymous>:11:396
```

I will replace that console warning with alert for further testing, I got a feeling its more common, but easily unnoticed in daily use.

**Comment by jdoe1024** [comment]
For https://www.shameless.com/dmca.php the problem comes from this uBlock rule (part of builtin uBlock filters):

    shameless.com##+js(aopr, document.dispatchEvent)

From what I understand from [uBlock wiki](https://github.com/gorhill/uBlock/wiki/Resources-Library#aoprjs-), this rule makes read access to `document.dispatchEvent()` throw a ReferenceError exception. The full stacktrace is as follow:

    Tampermonkey sandbox preparation failed. This usually is caused by a third-party extension. ReferenceError: wk5n4xl
        at HTMLDocument.abort (<anonymous>:9:15)
        at <anonymous>:9:2
        at Array.forEach (<anonymous>)
        at Object.d.createSafeDocument (<anonymous>:8:508)
        at <anonymous>:9:65
        at <anonymous>:9:88
        at <anonymous>:11:396
      y	@	VM43:1
      (anonymous)	@	VM43:9
      (anonymous)	@	VM43:11
      eval	@	common.js:2
      init	@	content.js:11
      (anonymous)	@	content.js:27
      (anonymous)	@	content.js:26
      (anonymous)	@	content.js:26
      (anonymous)	@	content.js:1
      b	@	content.js:1
      c	@	content.js:2
      (anonymous)	@	page.js:1

My guess is when TM inject its sandbox, it calls `document.dispatchEvent()` which uBlock blocks. This is probably more a uBlock issue than a TM one, but it would be great if both extension developers could get in touch to find a way to fix it (e.g., securely flagging TM scripts so that uBlock stops analyzing them).

There are similar rules for vidlox.me:

    vidlox.*##+js(abort-current-inline-script.js, Math, zfgloaded)
    vidlox.*##+js(abort-on-property-read.js, decodeURIComponent)
    vidlox.*##+js(nowebrtc.js)
    vidlox.*##+js(set-constant.js, popns, true)
    vidlox.*##+js(set-constant.js, popjs, true)

The uBlock rules for mixdrop have been updated and site seems to work with TM now.

**Comment by jdoe1024** [comment]
As a workaround, it is possible to disable uBlock scriptlet injection for those sites by adding the following filters (in "My filters"):

    mixdrop.co#@#+js()
    vidlox.*#@#+js()
    shameless.com#@#+js()

TamperMonkey always works with these rules, but this might let some annoying ads go through uBlock filtering...

**Comment by derjanb** [comment]
> For https://www.shameless.com/dmca.php the problem comes from this uBlock rule (part of builtin uBlock filters):

Nice find. I'll add a workaround like I did for `decodeURIComponent`.

**Comment by raszpl** [comment]
Nice one jdoe1024!
Is there a way or ordering extension execution in Chrome? That would seem the most future proof fix.

**Comment by jdoe1024** [comment]
> > For https://www.shameless.com/dmca.php the problem comes from this uBlock rule (part of builtin uBlock filters):
> 
> Nice find. I'll add a workaround like I did for `decodeURIComponent`.

I'm not sure what you mean here, but that workaround does not seem to work as `decodeURIComponent` is blocked by uBlock on https://vidlox.me/. Following are screenshots of the line in TM sandbox script that triggers the `ReferenceError` for all the 3 sites:
![image](https://user-images.githubusercontent.com/34872417/72285645-d99ee780-3643-11ea-90c7-1de36010279f.png)
![image](https://user-images.githubusercontent.com/34872417/72285660-defc3200-3643-11ea-94b1-c2323c4a261a.png)
![image](https://user-images.githubusercontent.com/34872417/72285679-e7ed0380-3643-11ea-843a-98f4c3400374.png)

They are all related to TM sandbox script accessing properties/methods blocked by uBlock filters.

**Comment by derjanb** [comment]
> I'm not sure what you mean here, but that workaround does not seem to work as decodeURIComponent is blocked by uBlock

I take everything back and claim the opposite. :-D

Based on my fix for https://github.com/Tampermonkey/tampermonkey/issues/637 I thought that I could fix it that way as well, but after further investigation I found, that I can only workaround `abort-current-inline-script.js` but not `abort-on-property-read.js`.

**Comment by derjanb** [comment]
I'll close this issue as "wontfix" even though it's a "cantfix". :-)

**Comment by jdoe1024** [comment]
You can workaround `shameless.com##+js(aopr, document.dispatchEvent)` using something like that:
```
var c = document.__proto__['dispatchEvent'].bind(document);
```

And you can also workaround `mixdrop.*##+js(acis, String.fromCharCode)` the same way you did for #637.
However, I see no workaround for `vidlox.*##+js(abort-on-property-read.js, decodeURIComponent)` since that method is only available in the window instance (not its prototype or anywhere else).

But you shouldn't have to workaround all these anyway, because there always be new scriptlet injection filters that will break TM sooner or later. The only sustainable solution would be for uBlock to stop messing up with injected script from other extensions. uBlock dev does not think [this is possible](https://github.com/uBlockOrigin/uAssets/issues/6827#issuecomment-574164425) and that filters that interfere with other extension should be reported so they might be fixed (but hope is low if I understand [this response](https://github.com/uBlockOrigin/uAssets/issues/6827#issuecomment-574186923)).
I do have a solution in mind that I'll try to propose to uBlock, but I doubt they'll accept it...

**Comment by raszpl** [comment]
Well, this will be tedious :o :)
Looks like I will be leaving my `{alert("sandbox preparation failed");a.console.error("Tampermonkey sandbox preparation failed. This usually is caused by a third-party extension.",b)}` in `page.js` (+ removing file write permission so Chrome doesnt "fix" it from under me) to see what needs adhoc interventions

```
vidlox.*#@#+js(aopr, decodeURIComponent)
mixdrop.*#@#+js(acis, String.fromCharCode)
```
in uBlock My filters fixes those particular two.

**Comment by lainverse** [comment]
Is there a way for extensions to communicate with each other? It might be possible to contact uBO author and develop with him some API to negotiate password between extensions and make properties work as expected when negotiated password is sent as extra parameter at the end.

**Comment by raszpl** [comment]
Thats too much work, especially considering uBO is on borrowed time anyway - Google is in the process of removing functionality necessary for uBO to function. We might not have uBO at all in 6-12 months.
